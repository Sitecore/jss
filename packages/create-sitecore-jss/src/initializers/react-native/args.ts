import { BaseArgs, ClientAppAnswer } from '../../common';

export type ReactNativeArgs = BaseArgs & Partial<Omit<ClientAppAnswer, 'fetchWith'>>;
