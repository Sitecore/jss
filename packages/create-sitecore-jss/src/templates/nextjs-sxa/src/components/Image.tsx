import React from 'react';
import {
  Image as JssImage,
  Link as JssLink,
  ImageField,
  Field,
  LinkField,
  Text,
  useSitecoreContext,
} from '@sitecore-jss/sitecore-jss-nextjs';

interface Fields {
  Image: ImageField;
  ImageCaption: Field<string>;
  TargetUrl: LinkField;
}

type ImageProps = {
  params: { [key: string]: string };
  fields: Fields;
};

const ImageDefault = (props: ImageProps): JSX.Element => (
  <div className={`component image ${props.params.styles}`.trimEnd()}>
    <div className="component-content">
      <span className="is-empty-hint">Image</span>
    </div>
  </div>
);

export const Default = (props: ImageProps): JSX.Element => {
  const { sitecoreContext } = useSitecoreContext();

  if (props.fields) {
    const Image = () => <JssImage field={props.fields.Image} />;

    return (
      <div className={`component image ${props.params.styles}`}>
        <div className="component-content">
          {sitecoreContext.pageState === 'edit' ? (
            <Image />
          ) : (
            <JssLink field={props.fields.TargetUrl}>
              <Image />
            </JssLink>
          )}
          <Text
            tag="span"
            className="image-caption field-imagecaption"
            field={props.fields.ImageCaption}
          />
        </div>
      </div>
    );
  }

  return <ImageDefault {...props} />;
};
