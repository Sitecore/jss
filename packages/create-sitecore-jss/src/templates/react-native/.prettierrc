{
  "endOfLine": "crlf",
  "singleQuote": true,
  "trailingComma": "es5",
  "tabWidth": 2,
  "arrowParens": "always",
  "printWidth": 100
}
