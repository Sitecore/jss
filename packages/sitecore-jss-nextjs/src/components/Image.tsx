import { mediaApi } from '@sitecore-jss/sitecore-jss/media';
import PropTypes from 'prop-types';
import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { convertAttributesToReactProps } from '@sitecore-jss/sitecore-jss-react';
import {
  default as NextImage,
  ImageLoader,
  ImageLoaderProps,
  ImageProps as NextImageProps,
} from 'next/image';
// import { getPublicUrl } from '../utils';
export interface ImageFieldValue {
  [attributeName: string]: unknown;
  src?: string;
  /** HTML attributes that will be appended to the rendered <img /> tag. */
}

export interface ImageField {
  value?: ImageFieldValue;
  editable?: string;
}

export interface ImageSizeParameters {
  [attr: string]: string | number | undefined;
  /** Fixed width of the image */
  w?: number;
  /** Fixed height of the image */
  h?: number;
  /** Max width of the image */
  mw?: number;
  /** Max height of the image */
  mh?: number;
  /** Ignore aspect ratio */
  iar?: 1 | 0;
  /** Allow stretch */
  as?: 1 | 0;
  /** Image scale. Defaults to 1.0 */
  sc?: number;
}

export interface ImageProps extends NextImageProps {
  [attributeName: string]: unknown;
  /** Image field data (consistent with other field types) */
  field?: ImageField | ImageFieldValue;

  /**
   * Can be used to explicitly disable inline editing.
   * If true and `media.editable` has a value, then `media.editable` will be processed
   * and rendered as component output. If false, `media.editable` value will be ignored and not rendered.
   */
  editable?: boolean;

  /**
   * Parameters that will be attached to Sitecore media URLs
   */
  imageParams?: {
    [paramName: string]: string | number;
  };

  /**
   * Custom regexp that finds media URL prefix that will be replaced by `/-/jssmedia` or `/~/jssmedia`.
   * @example
   * /\/([-~]{1})assets\//i
   * /-assets/website -> /-/jssmedia/website
   * /~assets/website -> /~/jssmedia/website
   */
  mediaUrlPrefix?: RegExp;

  /** HTML attributes that will be appended to the rendered <img /> tag. */
}

const getEditableWrapper = (editableMarkup: string, ...otherProps: unknown[]) => (
  // create an inline wrapper and use dangerouslySetInnerHTML.
  // if we try to parse the EE value, the parser will strip invalid or disallowed attributes from html elements - and EE uses several
  <span
    className="sc-image-wrapper"
    {...otherProps}
    dangerouslySetInnerHTML={{ __html: editableMarkup }}
  />
);

export const Image: React.SFC<ImageProps> = ({
  editable,
  imageParams,
  field,
  mediaUrlPrefix,
  ...otherProps
}) => {
  // next handles srcSet, throw error if srcSet is present
  if (otherProps.srcSet) {
    throw new Error(
      'srcSet not supported on Nextjs Image component. Use deviceSizes in nextjs.config: https://nextjs.org/docs/api-reference/next/image#device-sizes'
    );
  }

  // next handles src and we use a custom loader,
  // throw error if these are present
  if (otherProps.src || otherProps.loader) {
    // TODO: Refine error message
    throw new Error(
      'Detected conflicting props src or loader. If you wish to use these props, use next/image directly.'
    );
  }

  const dynamicMedia = field as ImageField | ImageFieldValue;

  if (
    !field ||
    (!dynamicMedia.editable && !dynamicMedia.value && !(dynamicMedia as ImageFieldValue).src)
  ) {
    return null;
  }

  const imageField = dynamicMedia as ImageField;

  // we likely have an experience editor value, should be a string
  if (editable && imageField.editable) {
    const foundImg = mediaApi.findEditorImageTag(imageField.editable);
    if (!foundImg) {
      return getEditableWrapper(imageField.editable);
    }

    const foundImgProps = convertAttributesToReactProps(foundImg.attrs);
    // Note: otherProps may override values from foundImgProps, e.g. `style`, `className` prop
    // We do not attempt to merge.
    const imgAttrs = { ...foundImgProps, ...otherProps };

    if (!imgAttrs) {
      return getEditableWrapper(imageField.editable);
    }

    const imgHtml = ReactDOMServer.renderToStaticMarkup(<img {...imgAttrs} />);
    const editableMarkup = imageField.editable.replace(foundImg.imgTag, imgHtml);
    return getEditableWrapper(editableMarkup);
  }

  // some wise-guy/gal is passing in a 'raw' image object value
  const img = (dynamicMedia as ImageFieldValue).src
    ? field
    : (dynamicMedia.value as ImageFieldValue);
  if (!img) {
    return null;
  }
  // this prop is needed for non-static images - set it to 1x1 transparent pixel base64 encoded if not supplied by user.
  if (!otherProps.blurDataURL) {
    otherProps.blurDataURL =
      'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=';
  }
  const attrs = { ...img, ...otherProps };

  if (attrs) {
    attrs.src = img;
    // TODO: Finalize loader for XM - export it? we need imageParams but also would
    // like the decouple this function.
    const xmLoader: ImageLoader = ({ src, width }: ImageLoaderProps): string => {
      const loaderProps = { mw: width };
      const mergedProps = { ...imageParams, ...loaderProps };
      // TODO:
      // hardcoded hostname at the moment to get around a  bug.
      // image loaders inside Next's repo like Cloudinary
      // have access to a root prop. We want to access root also if we need to inject the hostname.
      return `$https://cm.jss.localhost${mediaApi.updateImageUrl(
        src,
        mergedProps,
        mediaUrlPrefix
      )}`;
    };
    // TODO?: Create a loader for edge
    // TODO: export - do we need to do anything special for it? (we don't think so)

    return <NextImage loader={xmLoader} {...attrs} />;
  }

  return null; // we can't handle the truth
};

Image.propTypes = {
  field: PropTypes.oneOfType([
    PropTypes.shape({
      src: PropTypes.string.isRequired,
    }),
    PropTypes.shape({
      value: PropTypes.object,
      editable: PropTypes.string,
    }),
  ]),
  editable: PropTypes.bool,
  mediaUrlPrefix: PropTypes.instanceOf(RegExp),
  imageParams: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.number.isRequired, PropTypes.string.isRequired]).isRequired
  ),
};

Image.defaultProps = {
  editable: true,
};

Image.displayName = 'Image';
