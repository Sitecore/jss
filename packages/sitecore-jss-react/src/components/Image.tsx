import { mediaApi } from '@sitecore-jss/sitecore-jss';
import PropTypes from 'prop-types';
import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { convertAttributesToReactProps } from '../utils';

export interface ImageFieldValue {
  [attributeName: string]: unknown;
  src?: string;
  /** HTML attributes that will be appended to the rendered <img /> tag. */
}

export interface ImageField {
  value?: ImageFieldValue;
  editable?: string;
}

export interface ImageSizeParameters {
  [attr: string]: string | number | undefined;
  /** Fixed width of the image */
  w?: number;
  /** Fixed height of the image */
  h?: number;
  /** Max width of the image */
  mw?: number;
  /** Max height of the image */
  mh?: number;
  /** Ignore aspect ratio */
  iar?: 1 | 0;
  /** Allow stretch */
  as?: 1 | 0;
  /** Image scale. Defaults to 1.0 */
  sc?: number;
}

export interface ImageProps {
  [attributeName: string]: unknown;
  /**
   * The image field data.
   * @deprecated use field property instead
   */
  media?: ImageField | ImageFieldValue;

  /** Image field data (consistent with other field types) */
  field?: ImageField | ImageFieldValue;

  /**
   * Can be used to explicitly disable inline editing.
   * If true and `media.editable` has a value, then `media.editable` will be processed
   * and rendered as component output. If false, `media.editable` value will be ignored and not rendered.
   */
  editable?: boolean;

  /**
   * Parameters that will be attached to Sitecore media URLs
   */
  imageParams?: {
    [paramName: string]: string | number;
  };

  srcSet?: ImageSizeParameters[];

  /**
   * Custom regexp that finds media URL prefix that will be replaced by `/-/jssmedia` or `/~/jssmedia`.
   * @example
   * /\/([-~]{1})assets\//i
   * /-assets/website -> /-/jssmedia/website
   * /~assets/website -> /~/jssmedia/website
   */
  mediaUrlPrefix?: RegExp;

  /** HTML attributes that will be appended to the rendered <img /> tag. */
}

const getEditableWrapper = (editableMarkup: string, ...otherProps: unknown[]) => (
  // create an inline wrapper and use dangerouslySetInnerHTML.
  // if we try to parse the EE value, the parser will strip invalid or disallowed attributes from html elements - and EE uses several
  <span
    className="sc-image-wrapper"
    {...otherProps}
    dangerouslySetInnerHTML={{ __html: editableMarkup }}
  />
);

const getImageAttrs = (
  {
    src,
    srcSet,
    ...otherAttrs
  }: {
    [key: string]: unknown;
    src?: string;
    srcSet?: ImageSizeParameters[];
  },
  imageParams?: { [paramName: string]: string | number },
  mediaUrlPrefix?: RegExp
) => {
  if (!src) {
    return null;
  }

  const newAttrs: { [attr: string]: unknown } = {
    ...otherAttrs,
  };

  // update image URL for jss handler and image rendering params
  const resolvedSrc = mediaApi.updateImageUrl(src, imageParams, mediaUrlPrefix);
  if (srcSet) {
    // replace with HTML-formatted srcset, including updated image URLs
    newAttrs.srcSet = mediaApi.getSrcSet(resolvedSrc, srcSet, imageParams, mediaUrlPrefix);
  }
  // always output original src as fallback for older browsers
  newAttrs.src = resolvedSrc;
  return newAttrs;
};

export const Image: React.SFC<ImageProps> = ({
  media,
  editable,
  imageParams,
  field,
  mediaUrlPrefix,
  ...otherProps
}) => {
  // allows the mistake of using 'field' prop instead of 'media' (consistent with other helpers)
  if (field && !media) {
    media = field;
  }

  const dynamicMedia = media as ImageField | ImageFieldValue;

  if (
    !media ||
    (!dynamicMedia.editable && !dynamicMedia.value && !(dynamicMedia as ImageFieldValue).src)
  ) {
    return null;
  }

  const imageField = dynamicMedia as ImageField;

  // we likely have an experience editor value, should be a string
  if (editable && imageField.editable) {
    const foundImg = mediaApi.findEditorImageTag(imageField.editable);
    if (!foundImg) {
      return getEditableWrapper(imageField.editable);
    }

    const foundImgProps = convertAttributesToReactProps(foundImg.attrs);
    // Note: otherProps may override values from foundImgProps, e.g. `style`, `className` prop
    // We do not attempt to merge.
    const imgAttrs = getImageAttrs(
      { ...foundImgProps, ...otherProps },
      imageParams,
      mediaUrlPrefix
    );
    if (!imgAttrs) {
      return getEditableWrapper(imageField.editable);
    }

    const imgHtml = ReactDOMServer.renderToStaticMarkup(<img {...imgAttrs} />);
    const editableMarkup = imageField.editable.replace(foundImg.imgTag, imgHtml);
    return getEditableWrapper(editableMarkup);
  }

  // some wise-guy/gal is passing in a 'raw' image object value
  const img = (dynamicMedia as ImageFieldValue).src
    ? media
    : (dynamicMedia.value as ImageFieldValue);
  if (!img) {
    return null;
  }

  const attrs = getImageAttrs({ ...img, ...otherProps }, imageParams, mediaUrlPrefix);
  if (attrs) {
    return <img {...attrs} />;
  }

  return null; // we can't handle the truth
};

Image.propTypes = {
  media: PropTypes.oneOfType([
    PropTypes.shape({
      src: PropTypes.string,
    }),
    PropTypes.shape({
      value: PropTypes.object,
      editable: PropTypes.string,
    }),
  ]),
  field: PropTypes.oneOfType([
    PropTypes.shape({
      src: PropTypes.string,
    }),
    PropTypes.shape({
      value: PropTypes.object,
      editable: PropTypes.string,
    }),
  ]),
  editable: PropTypes.bool,
  mediaUrlPrefix: PropTypes.instanceOf(RegExp),
  imageParams: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.number.isRequired, PropTypes.string.isRequired]).isRequired
  ),
};

Image.defaultProps = {
  editable: true,
};

Image.displayName = 'Image';
