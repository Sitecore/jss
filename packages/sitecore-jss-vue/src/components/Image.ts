import { mediaApi } from '@sitecore-jss/sitecore-jss';
import { CreateElement, FunctionalComponentOptions, RenderContext } from 'vue';
import { generateHtmlTag } from '../utils';

export interface ImageFieldValue {
  [attributeName: string]: any;
  src?: string;
  /** HTML attributes that will be appended to the rendered <img /> tag. */
}

export interface ImageField {
  value?: ImageFieldValue;
  editable?: string;
}

export interface ImageProps {
  /** The image field data. */
  media: ImageField | ImageFieldValue;

  /**
   * Can be used to explicitly disable inline editing.
   * If true and `media.editable` has a value, then `media.editable` will be processed
   * and rendered as component output. If false, `media.editable` value will be ignored and not rendered.
   */
  editable?: boolean;

  /**
   * Custom regexp that finds media URL prefix that will be replaced by `/-/jssmedia` or `/~/jssmedia`.
   * @example
   * /\/([-~]{1})assets\//i
   * /-assets/website -> /-/jssmedia/website
   * /~assets/website -> /~/jssmedia/website
   */
  mediaUrlPrefix?: RegExp;

  /**
   * Parameters that will be attached to Sitecore media URLs
   */
  imageParams?: {
    [paramName: string]: string;
  };
}

const getEditableWrapper = (editableMarkup: string, createElement: CreateElement) =>
  // create an inline wrapper and use dangerouslySetInnerHTML.
  // if we try to parse the EE value, the parser may strip invalid or disallowed attributes from html elements - and EE uses several
  createElement('span', {
    class: 'sc-image-wrapper',
    domProps: { innerHTML: editableMarkup },
  });

const getImageAttrs = (
  {
    src,
    srcSet,
    ...otherAttrs
  }: {
    src?: string;
    srcSet?: any;
    otherAttrs?: any;
  },
  imageParams: any,
  mediaUrlPrefix?: RegExp
) => {
  if (!src) {
    return null;
  }
  const newAttrs: any = {
    ...otherAttrs,
  };

  // update image URL for jss handler and image rendering params
  const resolvedSrc = mediaApi.updateImageUrl(src, imageParams, mediaUrlPrefix);
  if (srcSet) {
    // replace with HTML-formatted srcset, including updated image URLs
    newAttrs.srcSet = mediaApi.getSrcSet(resolvedSrc, srcSet, imageParams, mediaUrlPrefix);
  } else {
    newAttrs.src = resolvedSrc;
  }
  return newAttrs;
};

// custom comment to enable eslint vue plugin for Vue components as objects: https://github.com/vuejs/eslint-plugin-vue#attention
// @vue/component
export const Image: FunctionalComponentOptions<ImageProps> = {
  functional: true,
  props: {
    media: { type: Object, required: true },
    editable: { type: Boolean, default: true },
    imageParams: { type: Object, default: () => ({}) },
    mediaUrlPrefix: { type: RegExp, default: undefined },
  },
  // Need to assign `any` return type because Vue type definitions are inaccurate.
  // The Vue type definitions set `render` to a return type of VNode and that's it.
  // However, it is possible to return null | string | VNode[] | VNodeChildrenArrayContents.
  render(createElement: CreateElement, context: RenderContext): any {
    const { media, editable, imageParams, mediaUrlPrefix } = context.props;
    const contextAttrs = context.data.attrs;

    if (!media || (!media.editable && !media.value && !media.src)) {
      return null;
    }

    // we likely have an experience editor value, should be a string
    if (editable && media.editable) {
      const foundImg = mediaApi.findEditorImageTag(media.editable);
      if (!foundImg) {
        return getEditableWrapper(media.editable, createElement);
      }

      const imgAttrs = getImageAttrs(
        { ...foundImg.attrs, ...contextAttrs },
        imageParams,
        mediaUrlPrefix
      );
      if (!imgAttrs) {
        return getEditableWrapper(media.editable, createElement);
      }

      const imgHtml = generateHtmlTag('img', imgAttrs);
      const editableMarkup = media.editable.replace(foundImg.imgTag, imgHtml);
      return getEditableWrapper(editableMarkup, createElement);
    }

    // some wise-guy/gal is passing in a 'raw' image object value
    const img = media.src ? media : media.value;
    if (!img) {
      return null;
    }

    const attrs = getImageAttrs({ ...img, ...contextAttrs }, imageParams, mediaUrlPrefix);
    if (attrs) {
      // in functional components, context.data should be passed along to the
      // `createElement` function in order to retain attributes and events
      // https://vuejs.org/v2/guide/render-function.html#Passing-Attributes-and-Events-to-Child-Elements-Components
      const data = { ...context.data, attrs };
      return createElement('img', data);
    }

    return null; // we can't handle the truth
  },
};
