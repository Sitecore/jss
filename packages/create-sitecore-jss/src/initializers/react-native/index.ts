import path from 'path';
import { prompt } from 'inquirer';
import { prompts, ReactNativeAnswer } from './prompts';
import { Initializer } from '../../common/Initializer';
import { transform } from '../../common/steps';
import { ReactNativeArgs } from './args';

export default class ReactNativeInitializer implements Initializer {
  get isBase() {
    return true;
  }

  async init(args: ReactNativeArgs) {
    const defaults = args.yes
      ? {
          appName: 'sitecore-jss-react-native',
          hostName: 'sitecore-jss-react-native.dev.local',
          language: '',
        }
      : {};

    const answers = await prompt<ReactNativeAnswer>(prompts, { ...defaults, ...args });

    const mergedArgs = {
      ...args,
      ...answers,
    };

    const templatePath = path.resolve(__dirname, '../../templates/react-native');
    await transform(templatePath, mergedArgs, {
      filter: (filePath: string) =>
        !!mergedArgs.language || !filePath.endsWith('{{language}}.json'),
    });

    const response = {
      appName: answers.appName,
    };

    return response;
  }
}
