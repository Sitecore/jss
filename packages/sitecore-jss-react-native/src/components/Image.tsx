import React from 'react';
import PropTypes from 'prop-types';
import { Image as NativeImage, ImageSourcePropType } from 'react-native';
import SvgUri from 'react-native-svg-uri';
import { mediaApi } from '@sitecore-jss/sitecore-jss';

export interface ImageFieldValue {
  /** HTML attributes that will be appended to the rendered <img /> tag. */
  [attributeName: string]: unknown;
  src?: string | number;
  width?: number;
  height?: number;
  style?: unknown;
}

export interface ImageField {
  value?: ImageFieldValue;
  editable?: string;
}

export interface ImageProps {
  /** HTML attributes that will be appended to the rendered <img /> tag. */

  [attributeName: string]: unknown;
  /** The image field data. */
  media: ImageField | ImageFieldValue | null;
  field?: ImageField | ImageFieldValue | null;

  /**
   * Parameters that will be attached to Sitecore media URLs
   */
  imageUrlParams?: {
    [paramName: string]: string;
  } | null;
}

const getImageAttrs = (
  {
    src,
    width,
    height,
    style,
    ...otherAttrs
  }: {
    [attr: string]: unknown;
    src?: string | number;
    width?: number | undefined;
    height?: number | undefined;
    style?: unknown;
  },
  imageUrlParams?: { [paramName: string]: string } | null
) => {
  if (!src) {
    return null;
  }

  const newAttrs: { [key: string]: unknown } = { ...otherAttrs };

  // for network images, "width" and "height" are required as style properties, otherwise the image likely won't display.
  // for static images, "width" and "height" are not required, but if they are passed in then add them to the style prop.
  const imgStyles: { [attr: string]: unknown } = {};
  // react-native doesn't seem to like `&&` assignment, so use if statements instead
  if (width) {
    imgStyles.width = typeof width !== 'number' ? Number(width) : width;
  }
  if (height) {
    imgStyles.height = typeof height !== 'number' ? Number(height) : height;
  }

  // react native styles/StyleSheets can be merged by passing them as an array to the native Image component
  newAttrs.style = style ? [imgStyles, style] : imgStyles;

  // in react-native, you need to "import" static assets via `require` statements
  // when the packager builds your app, it (presumably) statically analyzes your code, extracts
  // the assets that are `require`d into an array/map, then assigns them a numerical value.
  if (typeof src === 'number') {
    newAttrs.source = src;
  } else {
    // assume we have a "network image", i.e. something loaded via http(s)
    // update image URL for jss handler and image rendering params
    const uri = mediaApi.updateImageUrl(src, imageUrlParams);
    // for network images, the "source" prop is an object with at least "uri" value, e.g. { uri: 'http://aviato.com/myimg.jpg' }
    newAttrs.source = { uri };
  }

  return newAttrs;
};

export const isSvgImage = (source: ImageSourcePropType) => {
  const isSvg = /\.svg($|\?|\&)/g;
  const src = NativeImage.resolveAssetSource(source);

  return src && isSvg.test(src.uri);
};

export const Image: React.SFC<ImageProps> = ({ media, imageUrlParams, field, ...otherProps }) => {
  let dynamicMedia = media;

  // allows the mistake of using 'field' prop instead of 'media' (consistent with other helpers)
  if (field && !media) {
    dynamicMedia = field;
  }

  if (!dynamicMedia || (!dynamicMedia.value && !(dynamicMedia as ImageFieldValue).src)) {
    return null;
  }

  // some wise-guy/gal might pass in a 'raw' image object value
  const img = (dynamicMedia as ImageFieldValue).src
    ? dynamicMedia
    : (dynamicMedia as ImageField).value;
  if (!img) {
    return null;
  }

  const attrs = getImageAttrs({ ...(img as ImageFieldValue), ...otherProps }, imageUrlParams);

  if (attrs && isSvgImage(attrs.source as ImageSourcePropType)) return <SvgUri {...attrs} />;

  if (attrs) return <NativeImage {...attrs} source={attrs.source as ImageSourcePropType} />;

  return null; // we can't handle the truth
};

Image.propTypes = {
  media: PropTypes.oneOfType([
    PropTypes.shape({
      src: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    }),
    PropTypes.shape({
      value: PropTypes.object,
    }),
  ]),
  imageUrlParams: PropTypes.any,
};

Image.displayName = 'Image';
