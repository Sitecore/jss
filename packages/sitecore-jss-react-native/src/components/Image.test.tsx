import React from 'react';
import renderer from 'react-test-renderer';
import { StyleSheet, ImageSourcePropType } from 'react-native';
import { Image, isSvgImage } from './Image';

// In react-native, you need to "import" static assets via `require` statements.
// When the packager builds your app, it statically analyzes your code, extracts the assets that are `require`d into an array/map, then assigns them a numerical value.
// Presumably, it then uses that value as an index to retrieve/render the proper static asset.
// Hence, the `staticImages` defined below have numeric values for their `src` prop for mocking.
// Naturally, this doesn't actually test the `require` behavior of the packager, but that should be reserved for integration tests.

const staticImages = {
  'test-image-1': {
    src: 1,
  },
};

const networkImages = {
  'test-image-1': {
    src: 'https://jssapp/-/media/myfile.ashx&w=180&h=360',
    width: 180,
    height: 360,
    alt: 'Logo',
  },
};

const testData = [
  { label: 'static images', data: staticImages },
  { label: 'network images', data: networkImages },
];

describe('<Image />', () => {
  testData.forEach((dataSet) => {
    describe(`with ${dataSet.label}`, () => {
      describe('with direct image object, no value/editable', () => {
        test('should render properly', () => {
          const props = {
            media: dataSet.data['test-image-1'],
            resizeMode: 'cover',
          };
          const rendered = renderer.create(<Image {...props} />);

          expect(rendered).toMatchSnapshot();
        });
      });

      describe('with value/editable image object', () => {
        test('should render properly', () => {
          const props = {
            media: {
              value: dataSet.data['test-image-1'],
              editable: 'bob loblaw',
            },
            resizeMode: 'cover',
          };
          const rendered = renderer.create(<Image {...props} />);

          expect(rendered).toMatchSnapshot();
        });
      });

      describe('with style prop', () => {
        describe('with width or height specified in field data', () => {
          test('should render merged styles', () => {
            const styles = StyleSheet.create({
              logoImage: {
                margin: 20,
              },
            });
            const props = {
              media: {
                value: {
                  ...dataSet.data['test-image-1'],
                  width: '180',
                  height: '360',
                },
                editable: 'bob loblaw',
              },
              style: styles.logoImage,
              resizeMode: 'cover',
            };
            const rendered = renderer.create(<Image {...props} />);

            expect(rendered).toMatchSnapshot();
          });
        });
      });
    });
  });

  test('isSvgImage', () => {
    const ut = (source: ImageSourcePropType, expected: boolean) =>
      expect(isSvgImage(source)).toEqual(expected);

    ut({ uri: './x.svg' }, true);
    ut({ uri: './x.png' }, false);
    ut({ uri: 'https://jssapp/-/media/myfile.svg&w=180' }, true);
    ut({ uri: 'https://jssapp/-/media/myfile.svg?w=180' }, true);
    ut({ uri: 'https://jssapp/-/media/myfile.png?w=180' }, false);
  });
});
