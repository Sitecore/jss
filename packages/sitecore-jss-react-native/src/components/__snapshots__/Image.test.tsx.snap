// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> with network images with direct image object, no value/editable should render properly 1`] = `
<Image
  alt="Logo"
  resizeMode="cover"
  source={
    Object {
      "uri": "https://jssapp/-/media/myfile.ashx&w=180&h=360",
    }
  }
  style={
    Object {
      "height": 360,
      "width": 180,
    }
  }
/>
`;

exports[`<Image /> with network images with style prop with width or height specified in field data should render merged styles 1`] = `
<Image
  alt="Logo"
  resizeMode="cover"
  source={
    Object {
      "uri": "https://jssapp/-/media/myfile.ashx&w=180&h=360",
    }
  }
  style={
    Array [
      Object {
        "height": 360,
        "width": 180,
      },
      Object {
        "margin": 20,
      },
    ]
  }
/>
`;

exports[`<Image /> with network images with value/editable image object should render properly 1`] = `
<Image
  alt="Logo"
  resizeMode="cover"
  source={
    Object {
      "uri": "https://jssapp/-/media/myfile.ashx&w=180&h=360",
    }
  }
  style={
    Object {
      "height": 360,
      "width": 180,
    }
  }
/>
`;

exports[`<Image /> with static images with direct image object, no value/editable should render properly 1`] = `
<Image
  resizeMode="cover"
  source={1}
  style={Object {}}
/>
`;

exports[`<Image /> with static images with style prop with width or height specified in field data should render merged styles 1`] = `
<Image
  resizeMode="cover"
  source={1}
  style={
    Array [
      Object {
        "height": 360,
        "width": 180,
      },
      Object {
        "margin": 20,
      },
    ]
  }
/>
`;

exports[`<Image /> with static images with value/editable image object should render properly 1`] = `
<Image
  resizeMode="cover"
  source={1}
  style={Object {}}
/>
`;
